pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/bloom/BloomIndexType.java

```diff
@@ -0,0 +1,109 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.local.segment.index.bloom;
+
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.pinot.segment.local.segment.creator.impl.bloom.OnHeapGuavaBloomFilterCreator;
+import org.apache.pinot.segment.local.segment.index.readers.bloom.BloomFilterReaderFactory;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.Constants;
+import org.apache.pinot.segment.spi.V1Constants;
+import org.apache.pinot.segment.spi.creator.IndexCreationContext;
+import org.apache.pinot.segment.spi.index.FieldIndexConfigs;
+import org.apache.pinot.segment.spi.index.IndexHandler;
+import org.apache.pinot.segment.spi.index.IndexReaderFactory;
+import org.apache.pinot.segment.spi.index.IndexType;
+import org.apache.pinot.segment.spi.index.creator.BloomFilterCreator;
+import org.apache.pinot.segment.spi.index.reader.BloomFilterReader;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.apache.pinot.spi.config.table.BloomFilterConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+
+public class BloomIndexType implements IndexType<BloomFilterConfig, BloomFilterReader, BloomFilterCreator> {
+  public static final BloomIndexType INSTANCE = new BloomIndexType();
+
+  @Override
+  public String getId() {
+    return "bloom_filter";
+  }
+
+  @Override
+  public Class<BloomFilterConfig> getIndexConfigClass() {
+    return BloomFilterConfig.class;
+  }
+
+  @Override
+  public BloomFilterConfig getDefaultConfig() {
+    return BloomFilterConfig.DISABLED;
+  }
+
+  @Override
+  public BloomFilterConfig getConfig(TableConfig tableConfig, Schema schema) {
+    throw new UnsupportedOperationException("To be implemented in a future PR");
+  }
+
+  @Override
+  public BloomFilterCreator createIndexCreator(IndexCreationContext context, BloomFilterConfig indexConfig) {
+    int cardinality = context.getCardinality();
+    if (cardinality == Constants.UNKNOWN_CARDINALITY) {
+      // This is when we're creating bloom filters for non dictionary encoded cols where exact cardinality is not
+      // known beforehand.
+      // Since this field is only used for the estimate cardinality, using total # of entries instead
+      // TODO (saurabh) Check if we can do a better estimate
+      cardinality = context.getTotalNumberOfEntries();
+    }
+    return new OnHeapGuavaBloomFilterCreator(context.getIndexDir(), context.getFieldSpec().getName(), cardinality,
+        indexConfig, context.getFieldSpec().getDataType());
+  }
+
+  @Override
+  public IndexReaderFactory<BloomFilterReader> getReaderFactory() {
+    return new IndexReaderFactory.Default<BloomFilterConfig, BloomFilterReader>() {
```

Comment:
Why do we need a new instance of the Factory on each call? Could this be a static Factory instance shared across calls?

Reply:
We don't _need_ to create a new instance each type, as they are stateless. It was a matter of simplicity in the code.

To be honest I don't think that really matters. This instance is stateless, is going to live a very short time and is going to be called once per segment and query. It is very probable that it is going to be optimized by the JVM and do not actually be allocated in the heap. And even if it does, it is going to very rapidly be collected by the young GC.

Given all of that, I would prefer to just let the code here in order to make it easier to read, but if there is a majority that think that it is better to catch it, I can modify it

Reply:
Not just the GC, I referred to [scalar replacement](https://shipilev.net/jvm/anatomy-quarks/18-scalar-replacement/), but I'll change it given that seems clear that it worries several commiters